modules/catalog/src/main/java/org/apache/ignite/internal/catalog/CatalogManagerImpl.java

```diff
@@ -330,11 +333,30 @@ public CompletableFuture<Void> execute(List<CatalogCommand> commands) {
         return saveUpdateAndWaitForActivation(new BulkUpdateProducer(List.copyOf(commands)));
     }
 
+    /**
+     * Cleanup outdated catalog versions, which can't be observed after given timestamp (inclusively), and compact underlying update log.
+     *
+     * @param timestamp Earliest observable timestamp.
+     * @return Operation future.
+     */
+    public CompletableFuture<Void> compactCatalog(long timestamp) {
+        Catalog catalog = catalogAt(timestamp);
+
+        truncateUpTo(catalog.version(), catalog.time());
+
+        return updateLog.saveSnapshot(new SnapshotUpdate(catalog.version(), new SnapshotEntry(catalog))).thenAccept(ignore -> {});
+    }
+
     private void registerCatalog(Catalog newCatalog) {
         catalogByVer.put(newCatalog.version(), newCatalog);
         catalogByTs.put(newCatalog.time(), newCatalog);
     }
 
+    private void truncateUpTo(int version, long time) {
```

Comment:
If we will accept an entire `Catalog`, then there will be no chance to provide inconsistent `version` and `time`. WDYT?